### PR TITLE
Handle multi-line return types and multi-line tuples

### LINKED ISSUE #409: panic on Option.unwrap() inside rewrite_fn_base
Tried on hyper's `src/lib.rs`, default config.

```
thread '<main>' panicked at 'called `Option::unwrap()` on a `None` value', ../src/libcore/option.rs:365
stack backtrace:
   1:     0x7fd9005ac940 - sys::backtrace::tracing::imp::write::he27ef2eb40e30753Plt
   2:     0x7fd9005b3215 - panicking::log_panic::_<closure>::closure.39293
   3:     0x7fd9005b2c61 - panicking::log_panic::h20a17353a34447751lx
   4:     0x7fd900577783 - sys_common::unwind::begin_unwind_inner::hf876a8bedc1a7d89gds
   5:     0x7fd9005780e8 - sys_common::unwind::begin_unwind_fmt::h69a2566a09092307mcs
   6:     0x7fd9005aa801 - rust_begin_unwind
   7:     0x7fd9005fda7f - panicking::panic_fmt::h4197e07a6c9f36b2dMK
   8:     0x7fd9005f7fb8 - panicking::panic::h14dc0798a39fe65fKKK
   9:     0x5592494a7091 - items::_<impl>::rewrite_fn_base::h245a9ccad07c83e71ge
  10:     0x55924949cfa1 - items::_<impl>::rewrite_fn::h86303d256eb5c6baPde
  11:     0x55924949cccc - visitor::_<impl>::visit_fn::hb60d971ace235580gId
  12:     0x55924949e630 - visit::walk_impl_item::walk_impl_item::h1012711238313890203
  13:     0x55924949d9b2 - visit::walk_item::walk_item::h9000815299871932106
  14:     0x559249499cf6 - visitor::_<impl>::visit_item::h29732b7a57b85ca3tKd
  15:     0x5592494ede0f - _<impl>::build_controller::_<closure>::closure.28224
  16:     0x7fd901ca669c - driver::compile_input::h7f85a10f04a664a63ba
  17:     0x7fd901e0126b - run_compiler::haabf54047e5bdb33wsc
  18:     0x5592494ef8a2 - format::heec407419e483e2dkSk
  19:     0x5592494f0236 - run::hfc58449b26d7a00e9Sk
  20:     0x55924945fc34 - main::h5ca3068261bd1bc80da
  21:     0x7fd9005b29a4 - sys_common::unwind::try::try_fn::h1831094503452742175
  22:     0x7fd9005aa668 - __rust_try
  23:     0x7fd9005b263f - rt::lang_start::h95fec6e24ac2c1799ix
  24:     0x7fd8ffc32ec4 - __libc_start_main
  25:     0x55924945e148 - <unknown>
  26:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by nrc**
Closes #409

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -475,19 +475,29 @@ impl Rewrite for ast::Ty {
...
+                    wrap_str(ret, context.config.max_width, width - 2, offset + 1)
                 } else {
-                    let rewrites: Option<Vec<_>>;
-                    rewrites = tup_ret.iter()
-                                      .map(|item| item.rewrite(context, width, offset))
-                                      .collect();
-
-                    try_opt!(rewrites).join(", ")
-                };
+                    let items: Vec<_> = tup_ret.iter()
```
Comment (marcusklaas) [inline src/types.rs]: We can collect into `Option<Vec<_>>` here and wrap that in a `try_opt!`.

File: src/types.rs [hunk 2]
```diff
@@ -475,19 +475,29 @@ impl Rewrite for ast::Ty {
...
+                                               })
+                                               .collect();
+
+                    if items.iter().any(|i| i.is_none()) {
+                        return None;
+                    }
 
-                let ret = format!("({})", inner);
-                wrap_str(ret, context.config.max_width, width, offset)
+                    let items = items.into_iter().filter_map(|i| i.map(|i| ListItem::from_str(i)));
```
Comment (marcusklaas) [inline src/types.rs]: Wouldn't `itemize_list` be nicer? We'd automatically collect comments, too!

File: src/types.rs [hunk 3]
```diff
@@ -475,19 +475,29 @@ impl Rewrite for ast::Ty {
...
+                    list_helper(items,
```
Comment (marcusklaas) [inline src/types.rs]: If we call `write_list` directly, we won't even have to allocate a vector.

File: src/types.rs [hunk 4]
```diff
@@ -475,19 +475,29 @@ impl Rewrite for ast::Ty {
...
+                                width - 2,
```
Comment (marcusklaas) [inline src/types.rs]: Nit: unchecked subtraction

File: src/types.rs (hunk 3 above)
Comment (nrc) [inline src/types.rs]: write_list takes a vec of items, so I don't see how this avoids allocation.

Comment (nrc) [inline src/types.rs]: write_list takes a vec of items, so I don't see how this avoids allocation.

Comment (nrc) [inline src/types.rs]: write_list takes a vec of items, so I don't see how this avoids allocation.

**Comment by marcusklaas** [comment]
There's still a few unchecked subtractions. With or without those fixed, r+.

**Pull request merged**